### PR TITLE
volumes: return better error messages for unsupported task drivers

### PR DESCRIPTION
**Pull request by tgross**
Fixes https://github.com/hashicorp/nomad/issues/7774 (partially https://github.com/hashicorp/nomad/issues/6664 as well)

When an allocation runs for a task driver that can't support volume mounts, the mounting will fail in a way that can be hard to understand. With host volumes this usually means failing silently, whereas with CSI the operator gets inscrutable internals exposed in the `nomad alloc status`.

This changeset adds a `MountConfig` field to the task driver [`Capabilities`](https://www.nomadproject.io/docs/internals/plugins/task-drivers/#capabilities-capabilities-error) response. We validate this when the `csi_hook` or `volume_hook` fires and return a user-friendly error.

Three notes:
* We don't currently have a way to get driver capabilities up to the server, except through attributes. Validating this when the user initially submits the jobspec would be even better than what we're doing here (and could be useful for all our other capabilities), but that's out of scope for this PR.
* The `MountConfig` enum starts with "supports all" in order to support community plugins in a backwards compatible way, rather than cutting them off from volume mounting unexpectedly.
* I haven't included an update to the internals docs. I'm going to do that in a separate PR so I can include the missing networking fields as well without cluttering this PR.

### CODE REVIEW

File: client/allocrunner/csi_hook.go [hunk 1]
```diff
@@ -88,7 +90,20 @@ func (c *csiHook) claimVolumesFromAlloc() (map[string]*volumeAndRequest, error)
...
 		if volumeRequest.Type == structs.VolumeTypeCSI {
+
+			for _, task := range tg.Tasks {
+				caps, err := c.ar.GetTaskDriverCapabilities(task.Name)
+				if err != nil {
+					return nil, fmt.Errorf("could not validate task driver capabilities: %v", err)
+				}
+
+				if caps.MountConfigs == drivers.MountConfigSupportNone {
+					return nil, fmt.Errorf("task driver does not support CSI")
```
Comment (notnoop) [inline client/allocrunner/csi_hook.go:103]: Should we indicate which task or the task driver names in these error messages?

The `volume_hook.go` references will include the task names when logged hopefully, but not sure about this one....

Comment (tgross) [reply client/allocrunner/csi_hook.go:103]: 👍 Definitely some room for improvement. Will fix.

**Comment by tgross** [comment]
Cherry-picked to 0.11.3 branch as 4b0fe71

**Comment by shantanugadgil** [comment]
hi, this error is still present in `0.12.0-beta1` with host volumes

```hcl
job "host_volume_1" {
  datacenters = ["dc1"]

  type = "batch"

  group "mygroup" {
    volume "myworkspace" {
      type      = "host"
      source    = "workspace"
      read_only = true
    }

    task "mytask" {
      driver = "raw_exec"

      volume_mount {
        volume      = "myworkspace"
        destination = "/myworkspace"
      }

      template {
        data = <<EOH
#!/bin/bash

set -u

id -a

hostname

env | sort

sleep 30
EOH

        destination = "local/runme.bash"
      }

      config {
        command = "/bin/bash"
        args    = ["-x", "local/runme.bash"]
      }
    }
  }
}
```

**Comment by tgross** [comment]
> hi, this error is still present in 0.12.0-beta1 with host volumes

Hi @shantanugadgil You _should_ be getting an error because the `raw_exec` driver has `MountConfigSupportNone`. If you're not getting an error to that effect, can you open a new issue with the error message you're getting?

**Comment by shantanugadgil** [comment]
@tgross the issue is that `plan` and `run` do not complain about it, and silently accept it.

The error is seen in the `status` command, but I was hoping that the plan/run would not let this proceed.

**Comment by tgross** [comment]
Yes, that's a known limitation. See the top of the PR:

> We don't currently have a way to get driver capabilities up to the server, except through attributes. Validating this when the user initially submits the jobspec would be even better than what we're doing here (and could be useful for all our other capabilities), but that's out of scope for this PR.

But if you want to open a new issue with that as a feature request, feel free!

**Pull request merged**